src/main/java/io/logz/jmx2logzio/listener/prometheus/PromHttpsSyncSender.kt

```diff
@@ -0,0 +1,120 @@
+import io.logz.sender.exceptions.LogzioServerErrorException
+import org.apache.http.HttpHeaders
+import org.xerial.snappy.Snappy
+import java.io.BufferedReader
+import java.io.IOException
+import java.io.InputStreamReader
+import java.net.HttpURLConnection
+
+class PromHttpsSyncSender(
+    private val configuration: HttpsRequestConfiguration,
+    private val reporter: SenderStatusReporter
+) {
+    fun sendToLogzio(timeSeriesList: List<Types.TimeSeries>) {
+        try {
+            val writeRequest = Remote.WriteRequest.newBuilder()
+                .addAllTimeseries(timeSeriesList)
+                .build()
+            val payload: ByteArray = Snappy.compress(writeRequest.toByteArray())
+            var currentRetrySleep = configuration.initialWaitBeforeRetryMS.toLong()
+
+            for (currTry in 1..configuration.maxRetriesAttempts) {
+                var responseCode = 0
+                var responseMessage = ""
+
+                try {
+                    val conn = sendRequest(payload)
+                    responseCode = conn.responseCode
+                    responseMessage = conn.responseMessage
+
+                    if (responseCode == HttpURLConnection.HTTP_OK) {
+                        reporter.info("Successfully sent bulk to logz.io, size: " + payload.size)
+                        break
+                    }
+
+                    if (!shouldRetry(responseCode, responseMessage, conn)) break
+
+                    handleRetry(currentRetrySleep, currTry, responseCode, responseMessage, null)
+                } catch (e: IOException) {
+                    reporter.error("Got IO exception - " + e.message)
+                    handleRetry(currentRetrySleep, currTry, responseCode, responseMessage, e)
+                }
+
+                currentRetrySleep *= 2
+            }
+        } catch (e: InterruptedException) {
+            reporter.info("Got interrupted exception")
+            Thread.currentThread().interrupt()
+        }
+    }
+
+    private fun sendRequest(payload: ByteArray): HttpURLConnection {
+        val conn = configuration.logzioListenerUrl.openConnection() as HttpURLConnection
+        conn.requestMethod = conn.requestMethod
+        conn.setRequestProperty(HttpHeaders.CONTENT_LENGTH, payload.size.toString())
+        conn.setRequestProperty(HttpHeaders.CONTENT_TYPE, "application/x-protobuf")
+        conn.setRequestProperty(HttpHeaders.CONTENT_ENCODING, "snappy")
+        conn.setRequestProperty(HttpHeaders.AUTHORIZATION, "Bearer ${configuration.logzioToken}")
+
+        conn.readTimeout = configuration.socketTimeout
+        conn.connectTimeout = configuration.connectTimeout
+        conn.doOutput = true
+        conn.doInput = true
+        conn.outputStream.write(payload)
+
+        return conn
+    }
+
+    private fun shouldRetry(responseCode: Int, responseMessage: String, conn: HttpURLConnection): Boolean {
+        return when(responseCode) {
+            HttpURLConnection.HTTP_BAD_REQUEST -> {
```

Comment:
Actually, reviewing the error codes we send, in any http BTW: The only retry is 500, all others should WARN to log and avoid retrying.

Reply:
fixed to only retry on 500, and for all other to report a warning and halt